### PR TITLE
Support request shell pty

### PR DESCRIPTION
**Pull request by eldipa**
## PR Summary

`pystassh` does not support a direct interaction with a `shell`, only through an `execute` method. It is enough for executing a series of commands but it will not work for interactive programs.

This PR adds support for requesting a shell and read/write to/from it, new methods of `Channel` and makes `Channel` accessible from `Session`.

## Additional Notes

This PR is a follow up of #1 and #2 . The PRs #1 and #2 should be reviewed and merged before this one. I'll do the proper rebase of this PR to make the peer-review much easier after that.

### CODE REVIEW

**Comment by julienc91** [comment]
@eldipa Could you rebase and resolve the conflicts?

**Comment by eldipa** [comment]
@julienc91 Rebased but I couldn't figure out why the coverage decreased. From the reports seems that I didn't cover the raise of two exceptions but I checked in my local env and I'm sure that I'm covering those.

Probably I'm not understanding  `codecov`'s reports correctly.

**Comment by eldipa** [comment]
@julienc91 rebased with the latest `main` branch. Only 2 commits left.

**Comment by julienc91** [comment]
@eldipa Thanks again :smiley: Don't worry about the coverage. I made a few changes on the tests to reduce the amount of monkeypatching, and I also made sure that libssh would be available when running the tests in the CI. I'll take care of what's left to rebase with this PR !

File: pystassh/channel.py [hunk 1]
```diff
@@ -57,6 +60,120 @@ def __enter__(self):
...
+                )
+
+        ret = api.Api.ssh_channel_request_shell(self._channel)
+        if ret != api.SSH_OK:
+            raise exceptions.ChannelException(
+                "Request a shell failed: {}".format(self.get_error_message())
+            )
+        self._shell_requested = True
+
+    def read_nonblocking(self, size, from_stderr=False):
```
Comment (julienc91) [inline pystassh/channel.py:82]: `read` and `read_nonblocking` could be simplified to avoid code duplication

File: pystassh/session.py [hunk 2]
```diff
@@ -51,6 +51,10 @@ def __init__(
         self._session = None
         self._channel = None
 
+    @property
```
Comment (julienc91) [inline pystassh/session.py:54]: I could not find the use for this propery, so I removed it

File: tests/test_channel.py [hunk 3]
```diff
@@ -168,6 +168,213 @@ def fake_result_init(self, channel, command):
...
+        "pystassh.api.Api.ssh_channel_new", lambda *_: "<channel object>"
+    )
+    monkeypatch.setattr(
+        "pystassh.api.Api.ssh_channel_open_session", lambda *_: pystassh.api.SSH_OK
+    )
+    monkeypatch.setattr(
+        "pystassh.channel.Channel._is_open", lambda self: bool(self._channel)
+    )
+
+    with pytest.raises(pystassh.exceptions.ChannelException) as err:
```
Comment (julienc91) [inline tests/test_channel.py:187]: This test and all the similar ones can be simplified like this:

```
with pytest.raises(pystassh.exceptsions.ChannelException, match="The channel is not open"):
    channel.request_shell()
```

**Review by julienc91 — APPROVED** [review]
I already rebased your commits and already merged them on the `main` branch, that's why I'll close this PR. Thanks a lot for your work!

File: pystassh/session.py (hunk 2 above)
Comment (julienc91) [reply pystassh/session.py:54]: My bad, now that I'm writing the doc for this feature, I see the point!

Comment (eldipa) [reply pystassh/session.py:54]: haha, yeup, it is the door to use channel's `read` and `write`

**Comment by julienc91** [comment]
@eldipa I just published version 1.2.0 with your changes and an updated documentation. Once again, thank you for your work on this :smiley:

**Comment by eldipa** [comment]
@julienc91 It was a pleasure!

**Pull request closed without merging**